lib/student_greeter.rb

```diff
@@ -0,0 +1,84 @@
+class StudentGreeter
+  def self.greet_all_ungreeted_students
+    new.greet_all_ungreeted_students
+  end
+
+  def initialize
+    @greeters = User.where(greeter: true)
+  end
+
+  def greet_all_ungreeted_students
+    courses = Course.strictly_current
+    courses.each do |course|
+      nonstudents_in_course = course.nonstudents
+      possible_greeters = @greeters & nonstudents_in_course
+      next if possible_greeters.empty?
+      greeter = possible_greeters[0]
+      course.students_without_nonstudents.each do |student|
+        greet_if_ungreeted(student, greeter)
+      end
+    end
+  end
+
+  private
+
+  def greet_if_ungreeted(student, greeter)
+    return if student.greeted
+    if talk_page_blank? student
+      greet(student, greeter)
+      return
+    end
+    return if a_greeter_already_posted?(student)
+    greet(student, greeter)
+  end
+
+  def talk_page_blank?(student)
+    Wiki.get_page_content(student.talk_page).nil?
+  end
+
+  def a_greeter_already_posted?(student)
+    contributor_ids = ids_of_contributors_to_page(student.talk_page)
+    return false if (@greeters.pluck(:id) & contributor_ids).empty?
+    # Mark student as greeted if a greeter has already edited their talk page
+    student.update_attributes(greeted: true)
+    true
+  end
+
+  def ids_of_contributors_to_page(page_title)
+    contributors_response = Wiki.query contributors_query(page_title)
+    # TODO: exception handling for unexpected response data
+    # currently, that will just cause a NoMethodError, which is okay but not
+    # optimal, because it will likely break a rake task.
+    contributors = contributors_response.data['pages'].values[0]['contributors']
+    contributor_ids = contributors.map { |user| user['userid'] }
+    contributor_ids
```

Comment:
Are you trying to stay under 100 chars for RuboCop? Couldn't you make this whole method two lines?

```
contrib = Wiki.query(contributors_query(page_title))
contrib.data['pages'].values.first['contributors'].map { |u| user['userid'] }
```

If your argument is that isn't clear, the method name is pretty descriptive, and it's a private method.